lib/Core/SQLiteBuildDB.cpp

```diff
@@ -640,14 +581,27 @@ class SQLiteBuildDB : public BuildDB {
 
   
 private:
+  /// Local cache of database DBKeyID (values) to engine KeyIDs
+  llvm::DenseMap<uint64_t, KeyID> engineKeyIDs;
+
+  /// Local cache of database engine KeyIDs to DBKeyIDs
+  llvm::DenseMap<KeyID, DBKeyID> dbKeyIDs;
+
+  /// Lookup or create a DBKeyID for a given engine KeyID
+  ///
+  /// This method is not thread-safe. The caller must protect access via the
+  /// dbMutex.
+  DBKeyID getKeyID(KeyID keyID, std::string *error_out) {
```

Comment:
I think this function would be a little more readable if you split out a function that actually did the database lookup or insert operation, and the helper that then wrote the result into the caches, if we didn't have a cache entry.